File: foreign/csharp/Iggy_SDK/Contracts/TopicOptions.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using Apache.Iggy.Headers;
using Apache.Iggy.IggyClient;

namespace Apache.Iggy.Contracts;

/// <summary>
///     Topic options that have no named parameter of their own, typed instead of keyed by hand.
/// </summary>
/// <remarks>
///     Hand the result of <see cref="ToDictionary" /> to the <c>options</c> parameter of
///     <see cref="IIggyTopic.CreateTopicAsync" />. Every key here is settable at creation only: an
///     update takes just the options that have a parameter of their own and refuses the rest by
///     name, the same way it refuses a key outside the server's catalog, so a mistyped key fails
///     the call rather than being ignored. <see cref="IIggySystem.DescribeOptionsAsync" />
///     enumerates the keys a given server accepts, with the kind and default of each.
///     A property left null emits no key, which leaves the server default in place.
/// </remarks>
public sealed class TopicOptions
{
    private const string SegmentSizeKey = "segment_size";
    private const string EnforceFsyncKey = "enforce_fsync";
    private const string MessagesRequiredToSaveKey = "messages_required_to_save";
    private const string SizeOfMessagesRequiredToSaveKey = "size_of_messages_required_to_save";
    private const string PreallocateSegmentsKey = "preallocate_segments";

    /// <summary>
    ///     Size of a single segment in bytes. The server checks it against its own bounds, a
    ///     multiple of 512 within the range it reports for the key.
    /// </summary>
    public ulong? SegmentSize { get; init; }

    /// <summary>
    ///     Whether writes to this topic's partitions are fsynced.
    /// </summary>
    public bool? EnforceFsync { get; init; }

    /// <summary>
    ///     Flush the journal once it holds this many messages. Must be non-zero.
    /// </summary>
    public uint? MessagesRequiredToSave { get; init; }

    /// <summary>
    ///     Flush the journal once it holds this many bytes. Paired with
    ///     <see cref="MessagesRequiredToSave" />: whichever threshold trips first flushes.
    /// </summary>
    public ulong? SizeOfMessagesRequiredToSave { get; init; }

    /// <summary>
    ///     Reserve a segment's bytes up front on a filesystem that supports it. Reserves exactly
    ///     <see cref="SegmentSize" />, so the two belong to one decision.
    /// </summary>
    public bool? PreallocateSegments { get; init; }

    /// <summary>
    ///     Renders the options that were set, each under the kind the server's catalog gives its key.
    /// </summary>
    /// <returns>Option values keyed by option name, empty when nothing was set.</returns>
    public Dictionary<string, HeaderValue> ToDictionary()
    {
        var options = new Dictionary<string, HeaderValue>();

        if (SegmentSize is { } segmentSize)
        {
            options[SegmentSizeKey] = HeaderValue.FromUInt64(segmentSize);
        }

        if (EnforceFsync is { } enforceFsync)
        {
            options[EnforceFsyncKey] = HeaderValue.FromBool(enforceFsync);
        }

        if (MessagesRequiredToSave is { } messagesRequiredToSave)
        {
            options[MessagesRequiredToSaveKey] = HeaderValue.FromUInt32(messagesRequiredToSave);
        }

        if (SizeOfMessagesRequiredToSave is { } sizeOfMessagesRequiredToSave)
        {
            options[SizeOfMessagesRequiredToSaveKey] = HeaderValue.FromUInt64(sizeOfMessagesRequiredToSave);
        }

        if (PreallocateSegments is { } preallocateSegments)
        {
            options[PreallocateSegmentsKey] = HeaderValue.FromBool(preallocateSegments);
        }

        return options;
    }
}

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/topic/TopicOptions.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iggy.topic;

import org.apache.iggy.message.HeaderValue;

import java.math.BigInteger;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Typed builder for the topic option keys that have no parameter of their own on
 * {@code createTopic}.
 *
 * <p>Compression, message expiry and max topic size are named parameters already. The five keys here
 * would otherwise have to be spelled as raw strings paired with hand-encoded values, where a typo in
 * the key or the wrong kind for the value is only caught by the server refusing the create.
 *
 * <p>These are create-time keys: an update refuses them by name, because they describe how a
 * partition's storage was laid down. {@code describeOptions} enumerates what a given server accepts.
 *
 * <pre>{@code
 * var options = TopicOptions.builder()
 *         .segmentSize(BigInteger.valueOf(134_217_728))
 *         .enforceFsync(true)
 *         .build();
 * topicsClient.createTopic(streamId, 1L, CompressionAlgorithm.None,
 *         BigInteger.ZERO, BigInteger.ZERO, "orders", options);
 * }</pre>
 */
public final class TopicOptions {

    private TopicOptions() {}

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {

        private final Map<String, HeaderValue> options = new LinkedHashMap<>();

        private Builder() {}

        /** Per-topic segment size in bytes: a 512-byte multiple within the server's bounds. */
        public Builder segmentSize(BigInteger bytes) {
            options.put("segment_size", HeaderValue.fromUint64(bytes));
            return this;
        }

        /** Whether writes to this topic's partitions fsync. */
        public Builder enforceFsync(boolean enabled) {
            options.put("enforce_fsync", HeaderValue.fromBool(enabled));
            return this;
        }

        /** Flush the journal once it holds this many messages. Must be non-zero. */
        public Builder messagesRequiredToSave(long messages) {
            options.put("messages_required_to_save", HeaderValue.fromUint32(messages));
            return this;
        }

        /** Flush the journal once it holds this many bytes. */
        public Builder sizeOfMessagesRequiredToSave(BigInteger bytes) {
            options.put("size_of_messages_required_to_save", HeaderValue.fromUint64(bytes));
            return this;
        }

        /**
         * Reserve each segment's bytes up front where the filesystem supports it.
         *
         * <p>The reservation is real disk and runs inline on the owning shard, so the server caps
         * {@code segment_size * partitions_count} at admission.
         */
        public Builder preallocateSegments(boolean enabled) {
            options.put("preallocate_segments", HeaderValue.fromBool(enabled));
            return this;
        }

        /**
         * The keys set on this builder, in the order they were set.
         *
         * <p>Wrapping a {@link LinkedHashMap} rather than calling {@code Map.copyOf}, whose
         * iteration order is salted per JVM run: the encoder walks this map in order, so a
         * randomized one would give the same builder a different block on every run.
         */
        public Map<String, HeaderValue> build() {
            return Collections.unmodifiableMap(new LinkedHashMap<>(options));
        }
    }
}
